datanode/api/trading_data_v2.go

```diff
@@ -2632,16 +2644,201 @@ func (t *TradingDataServiceV2) estimateMargin(
+		return nil, formatE(ErrMarketServiceGetMarketData, err)
+	}
+
+	mktProto := mkt.ToProto()
+
+	asset, err := mktProto.GetAsset()
+	if err != nil {
+		return nil, formatE(err)
+	}
+
+	marketObservable, err := num.DecimalFromString(mktData.MarkPrice.String())
+	if err != nil {
+		return nil, formatE(err)
+	}
+
+	auction := mktData.AuctionEnd > 0
+	if auction && mktData.MarketTradingMode == types.MarketTradingModeOpeningAuction.String() {
+		marketObservable = mktData.IndicativePrice
+	}
+
+	priceFactor, err := t.getMarketPriceFactor(ctx, mkt)
+	if err != nil {
+		return nil, err
+	}
+
+	dPriceFactor := priceFactor.ToDecimal()
+
+	uMarketObservable, overflowed := num.UintFromDecimal(marketObservable)
+	if overflowed {
+		return nil, formatE(fmt.Errorf("mark price overflowed: %s", mktData.MarkPrice))
+	}
+
+	marketObservableInAssetPrecision, err := t.scaleFromMarketToAssetPrice(ctx, mkt, uMarketObservable)
+	if err != nil {
+		return nil, formatE(ErrScalingPriceFromMarketToAsset, err)
+	}
+
+	dMarketObservableInAssetPrecision := num.DecimalFromUint(marketObservableInAssetPrecision)
+
+	positionFactor := num.DecimalFromFloat(10).
+		Pow(num.DecimalFromInt64(int64(mkt.PositionDecimalPlaces)))
+
+	linearSlippageFactor, err := num.DecimalFromString(mktProto.LinearSlippageFactor)
+	if err != nil {
+		return nil, formatE(fmt.Errorf("can't parse linear slippage factor: %s", mktProto.LinearSlippageFactor), err)
+	}
+	quadraticSlippageFactor, err := num.DecimalFromString(mktProto.QuadraticSlippageFactor)
+	if err != nil {
+		return nil, formatE(fmt.Errorf("can't parse quadratic slippage factor: %s", mktProto.QuadraticSlippageFactor), err)
+	}
+
+	buyOrders := make([]*risk.OrderInfo, 0, len(req.Orders))
+	sellOrders := make([]*risk.OrderInfo, 0, len(req.Orders))
+
+	for _, o := range req.Orders {
+		if o == nil {
+			continue
+		}
+		var price num.Decimal
+		p, err := num.DecimalFromString(o.Price)
+		if err != nil {
+			return nil, errors.Wrap(ErrInvalidOrderPrice, err.Error())
+		}
+		price = p.Mul(dPriceFactor)
+
+		if o.Side == types.SideBuy {
+			buyOrders = append(buyOrders, &risk.OrderInfo{Size: o.Remaining, Price: price, IsMarketOrder: o.IsMarketOrder})
+		}
+		if o.Side == types.SideSell {
+			sellOrders = append(sellOrders, &risk.OrderInfo{Size: o.Remaining, Price: price, IsMarketOrder: o.IsMarketOrder})
+		}
+	}
+
+	marginEstimate := t.computeMarginRange(
+		req.MarketId,
+		req.OpenVolume,
+		buyOrders,
+		sellOrders,
+		dMarketObservableInAssetPrecision,
+		positionFactor,
+		linearSlippageFactor,
+		quadraticSlippageFactor,
+		rf,
+		auction,
+		asset,
+		mkt.TradableInstrument.MarginCalculator.ScalingFactors)
+
+	var liquidationEstimate *v2.LiquidationEstimate
+	if req.CollateralAvailable != nil && len(*req.CollateralAvailable) > 0 {
+		collateralAvailable, err := num.DecimalFromString(*req.CollateralAvailable)
+		if err != nil {
+			return nil, formatE(fmt.Errorf("could not parse collateral available ('%s')", *req.CollateralAvailable), err)
```

Comment:
Could you make the first error  here a newInvalidArgumentError (look at the errors.go) it will map to the approriate error in the http response.
Also I'd add this validation at the beginning as it unecessary to do all the previous stuff if there's an user input.

Reply:
Done now